### PR TITLE
KeyVault Certificate client API review consistency update

### LINKED ISSUE #6790: Certificates consistency fixes
- [x] isEnabled is listed as a Boolean in sync beginCreateCertificate but is a boolean in async
- [x] includePending is a Boolean in async listPropertiesOfCertificates but a boolean in sync
- [x] certificateProperties parameter -> properties in updateCertificateProperties
- [x] (not 100% sure) get rid of the createIssuer overload that takes in issuerName and provider
- [ ] Make provider required for the CertificateIssuer model
- [ ] You are the only language that allows users to setProperties of KeyVaultCertificate. Should probably get trid
- [x] You are the only language that has a setter for CertificatePolicy in KeyVaultCertificateWithPolicy. Should probably get rid
- [x] Expose vaultUrl for CertificateProperties
- [x] Need constructor in CertificatePolicy that takes both subject and subjectAlternativeNames
- [x] Get rid of getSubjectName in CertificatePolicy
- [x] Inconsistency since CertificatePolicy constructor takes subject and subjectAlternativeNames, but there is no setSubject and there is setSubjectAlternativeNames
- [x] certificateName parameter -> name in MergeCertificateOptions
- [ ] You require at least firstName, lastName, and email for AdministratorContact. None of these are specifically required, just one out of the four are.
- [ ] Same thing for CertificateContact
- [x] getCreated() and getUpdated() -> getCreatedOn() and getUpdatedOn() in CertificateIssuer
- [x] Bigger discussion, but Java doesn't have cancelCertificateOperation on the client. Having it on the CertificateOperation itself requires a get and update permission. Is this something we're ok with?
- [x] What should you do about potentially not having get permissions in a poller? (will discuss as a team)

### PR DESCRIPTION
**Pull request by samvaity**
Fixes #6790 and [Certificate consistency meeting 3](https://microsoft.sharepoint.com/teams/AzureDeveloperExperience/_layouts/15/Doc.aspx?sourcedoc=%7Bd96e1ac2-3064-4c8f-87a8-69f575f99a84%7D&action=edit&wd=target%28Release.one%7C727d79ca-765a-4572-a0e9-bf7323fb398b%2FCertificates%20consistency%20meeting%203%7C88d10b80-647a-4ca2-aa1c-dd6c53d71735%2F%29&wdorigin=703&wdpreservelink=1)

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java [hunk 1]
```diff
@@ -445,6 +445,10 @@ public String getVaultUrl() {
                         return Mono.defer(() -> Mono.just(new PollResponse<>(LongRunningOperationStatus.IN_PROGRESS,
                             pollingContext.getLatestResponse().getValue())));
                     }
+                    if (deletedCertificateResponse.getStatusCode() == HttpURLConnection.HTTP_FORBIDDEN) {
+                        return Mono.defer(() -> Mono.just(new PollResponse<>(LongRunningOperationStatus.fromString("FORBIDDEN", true),
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java]: Based on the feedback from API review.
this is not supposed to be here.
This logic is supposed to be in the poll operation of beginCreateCertificate.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java [hunk 2]
```diff
@@ -612,6 +616,10 @@ public String getVaultUrl() {
                         return Mono.defer(() -> Mono.just(new PollResponse<>(LongRunningOperationStatus.IN_PROGRESS,
                             pollingContext.getLatestResponse().getValue())));
                     }
+                    if (certificateResponse.getStatusCode() == HttpURLConnection.HTTP_FORBIDDEN) {
+                        return Mono.defer(() -> Mono.just(new PollResponse<>(LongRunningOperationStatus.fromString("FORBIDDEN", true),
+                            pollingContext.getLatestResponse().getValue())));
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java]: Based on the feedback from API review.
this is not supposed to be here.
This logic is supposed to be in the poll operation of beginCreateCertificate.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateService.java [hunk 3]
```diff
@@ -111,8 +111,7 @@
                                                        Context context);
 
     @Get("certificates/{certificate-name}/{certificate-version}")
-    @ExpectedResponses({200, 404})
-    @UnexpectedResponseExceptionType(code = {403}, value = ResourceModifiedException.class)
+    @ExpectedResponses({200, 404, 403})
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateService.java]: we need to do this in beginCreateCertificate Polling operation service call, not here.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateService.java [hunk 4]
```diff
@@ -197,7 +196,7 @@
                                                    Context context);
 
     @Get("deletedcertificates/{certificate-name}")
-    @ExpectedResponses({200, 404})
+    @ExpectedResponses({200, 403, 404})
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateService.java]: we need to do this in beginCreateCertificate Polling operation service call, not here.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/AdministratorContact.java [hunk 5]
```diff
@@ -90,4 +65,44 @@ public String getEmail() {
...
+     */
+    public AdministratorContact setEmail(String email) {
+        this.email = email;
+        return this;
+    }
+
+    /**
+     * Set the contact of the admin.
+     * @param phone the enabled status to set
+     * @return the AdministratorContact object itself.
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/AdministratorContact.java]: the updated AdministratorContact object itself.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateOperation.java [hunk 6]
```diff
@@ -177,10 +190,45 @@ public String getRequestId() {
         return this.requestId;
     }
 
+    /**
+     * Get the URL for the Azure KeyVault service.
+     *
+     * @return the value of the URL for the Azure KeyVault service.
+     */
+    public String getVaultUrl() {
+        unpackId(this.id);
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateOperation.java]: this should get called at the time of deserialization.
Use jackson properties to call this at the time of deserialization. 
CertificateProperties does it as an examples.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateProperties.java [hunk 7]
```diff
@@ -141,6 +146,16 @@ public OffsetDateTime getUpdatedOn() {
         return this.tags;
     }
 
+    /**
+     * Get the URL for the Azure KeyVault service.
+     *
+     * @return the value of the URL for the Azure KeyVault service.
+     */
+    public String getVaultUrl() {
+        unpackId(this.id);
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateProperties.java]: same here.
This call is not needed.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateOperation.java [hunk 8]
```diff
@@ -177,10 +190,45 @@ public String getRequestId() {
...
+        return this.vaultUrl;
+    }
+
+    /**
+     * Get the certificate name.
+     *
+     * @return the name of the certificate.
+     */
+    public String getName() {
+        unpackId(this.id);
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateOperation.java]: same here.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/ImportCertificateOptions.java [hunk 9]
```diff
@@ -78,17 +78,17 @@ public Boolean isEnabled() {
...
 
     /**
      * Set the management policy for the certificate.
-     * @param certificatePolicy the management policy for the certificate
+     * @param policy the management policy for the certificate
      * @return the updated CertificateImportOptions itself
      */
-    public ImportCertificateOptions setCertificatePolicy(CertificatePolicy certificatePolicy) {
-        this.certificatePolicy = certificatePolicy;
+    public ImportCertificateOptions setPolicy(CertificatePolicy policy) {
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/ImportCertificateOptions.java]: Is this covered from the feedback?

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateWithPolicy.java [hunk 10]
```diff
@@ -43,8 +43,8 @@ public KeyVaultCertificateWithPolicy setProperties(CertificateProperties propert
      * Get the certificate policy of the certificate
      * @return the cer content.
      */
-    public CertificatePolicy getCertificatePolicy() {
-        return this.certificatePolicy;
+    public CertificatePolicy getPolicy() {
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateWithPolicy.java]: CertificatePolicy -> policy , covered from the feedback ?

File: sdk/keyvault/azure-security-keyvault-certificates/src/test/java/com/azure/security/keyvault/certificates/CertificateClientTest.java [hunk 11]
```diff
@@ -381,18 +381,6 @@ public void createIssuer() {
         });
     }
 
-    @Test
-    public void createIssuerEmptyName() {
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/test/java/com/azure/security/keyvault/certificates/CertificateClientTest.java]: we should port these tests to use CertificateIssuer

File: sdk/keyvault/azure-security-keyvault-certificates/src/test/java/com/azure/security/keyvault/certificates/CertificateClientTest.java [hunk 12]
```diff
@@ -381,18 +381,6 @@ public void createIssuer() {
...
-        assertRestException(() -> client.createIssuer("", ""),
-            HttpResponseException.class, HttpURLConnection.HTTP_BAD_METHOD);
-    }
-
-    @Test
-    public void createIssuerNullProvider() {
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/test/java/com/azure/security/keyvault/certificates/CertificateClientTest.java]: same here

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateIssuer.java [hunk 13]
```diff
@@ -85,6 +85,15 @@ public CertificateIssuer(String name, String provider) {
         this.provider = provider;
     }
 
+    /**
+     * Creates an instance of the issuer.
+     *
+     * @param name The name of the issuer.
+     */
+    public CertificateIssuer(String name) {
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateIssuer.java]: Do we need this ?
Can create confusion for user as name and provider to create a valid issuer on the service side.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/ImportCertificateOptions.java (hunk 9 above)
Comment (samvaity) [reply sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/ImportCertificateOptions.java]: yes, [here](onenote:https://microsoft.sharepoint.com/teams/AzureDeveloperExperience/Shared%20Documents/Service%20-%20KeyVault/Azure%20SDK%20-%20Service%20-KeyVault/Release.one#Certificates%20consistency%20meeting%203&section-id=727d79ca-765a-4572-a0e9-bf7323fb398b&page-id=88d10b80-647a-4ca2-aa1c-dd6c53d71735&end)

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateWithPolicy.java (hunk 10 above)
Comment (samvaity) [reply sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/KeyVaultCertificateWithPolicy.java]: yup, [here](https://microsoft.sharepoint.com/teams/AzureDeveloperExperience/Shared%20Documents/Service%20-%20KeyVault/Azure%20SDK%20-%20Service%20-KeyVault/Release.one#Certificates%20consistency%20meeting%203&section-id=727d79ca-765a-4572-a0e9-bf7323fb398b&page-id=88d10b80-647a-4ca2-aa1c-dd6c53d71735&end)

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateIssuer.java (hunk 13 above)
Comment (samvaity) [reply sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateIssuer.java]: But they are not required when you want to update so this 'ctor overload. Also, will mention in the methods where the provider should be required to clarify this.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java (hunk 1 above)
Comment (samvaity) [reply sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java]: As discussed offline, we should not be swallowing 403 on beginCreateCertificate but on beginRecover and beginDelete polling operations. And the poller should return a `SUCCESSFULLY_COMPLETED` poller operation.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateOperation.java [hunk 14]
```diff
@@ -177,10 +190,43 @@ public String getRequestId() {
...
+    void unpackId(String id) {
+        if (id != null && id.length() > 0) {
+            this.id = id;
+            try {
+                URL url = new URL(id);
+                String[] tokens = url.getPath().split("/");
+                this.vaultUrl = (tokens.length >= 1 ? tokens[1] : null);
+                this.name = (tokens.length >= 3 ? tokens[2] : null);
+            } catch (MalformedURLException e) {
+                e.printStackTrace();
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateOperation.java]: this should be logged properly.
in Certificate Properties too.

File: sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateProperties.java [hunk 15]
```diff
@@ -237,6 +251,7 @@ void unpackId(String id) {
             try {
                 URL url = new URL(id);
                 String[] tokens = url.getPath().split("/");
+                this.vaultUrl = (tokens.length >= 1 ? tokens[1] : null);
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/models/CertificateProperties.java]: tokens.length >= 2 ? .....

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md [hunk 16]
```diff
@@ -1,4 +1,13 @@
 # Release History
+## 4.0.0-beta.7 (2019-12-19)
+- `beginDeleteCertificate` and `beginRecoverDeletedCertificate` polling operation to be `SUCCESSFULLY_COMPLETED` on 403.
+- `CertificateClient.createIssuer` and `CertificateAsyncClient.createIssuer` now require a `CertificateIssuer` with both a name and provider.
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md]: mention that the overload of (name, provider) is removed

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md [hunk 17]
```diff
@@ -1,4 +1,13 @@
...
+- Removed `setProperties` from `KeyVaultCertificate` model.
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md]: I don't see this change in this PR.

File: sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md [hunk 18]
```diff
@@ -1,4 +1,13 @@
 # Release History
+## 4.0.0-beta.7 (2019-12-19)
+- `beginDeleteCertificate` and `beginRecoverDeletedCertificate` polling operation to be `SUCCESSFULLY_COMPLETED` on 403.
```
Comment (g2vinay) [inline sdk/keyvault/azure-security-keyvault-certificates/CHANGELOG.md]: `beginDeleteCertificate` and `beginRecoverDeletedCertificate` methods now return a poll response with a status of `SUCCESSFULLY_COMPLETED` when service returns 403 status.

**Pull request merged**